istioctl/cmd/revision.go

```diff
@@ -0,0 +1,558 @@
+	}
+	return w.Flush()
+}
+
+func getEnabledComponents(iops *v1alpha1.IstioOperatorSpec) []string {
+	enabledComponents := []string{}
+	if iops.Components.Base.Enabled.GetValue() {
+		enabledComponents = append(enabledComponents, "base")
+	}
+	if iops.Components.Cni.Enabled.GetValue() {
+		enabledComponents = append(enabledComponents, "cni")
+	}
+	if iops.Components.Pilot.Enabled.GetValue() {
+		enabledComponents = append(enabledComponents, "istiod")
+	}
+	if iops.Components.IstiodRemote.Enabled.GetValue() {
+		enabledComponents = append(enabledComponents, "istiod-remote")
+	}
+	for _, gw := range iops.Components.IngressGateways {
+		if gw.Enabled.Value {
+			enabledComponents = append(enabledComponents, fmt.Sprintf("ingress:%s", gw.GetName()))
+		}
+	}
+	for _, gw := range iops.Components.EgressGateways {
+		if gw.Enabled.Value {
+			enabledComponents = append(enabledComponents, fmt.Sprintf("egress:%s", gw.GetName()))
+		}
+	}
+	return enabledComponents
+}
+
+func getIOPs(restConfig *rest.Config) ([]*iopv1alpha1.IstioOperator, error) {
+	client, err := dynamic.NewForConfig(restConfig)
+	if err != nil {
+		return nil, err
+	}
+	ul, err := client.
+		Resource(istioOperatorGVR).
+		Namespace(istioNamespace).
+		List(context.TODO(), meta_v1.ListOptions{})
+	if err != nil {
+		return nil, err
+	}
+	iops := []*iopv1alpha1.IstioOperator{}
+	for _, un := range ul.Items {
+		un.SetCreationTimestamp(meta_v1.Time{}) // UnmarshalIstioOperator chokes on these
+		un.SetManagedFields([]meta_v1.ManagedFieldsEntry{})
+		by := util.ToYAML(un.Object)
+		iop, err := operator_istio.UnmarshalIstioOperator(by, true)
+		if err != nil {
+			return nil, err
+		}
+		iops = append(iops, iop)
+	}
+	return iops, nil
+}
+
+// TODO: Get Rid of magic strings or get rid of these methods altogether
+func getControlPlanePods(restConfig *rest.Config, revision string) ([]v1.Pod, error) {
+	return getPodsForComponent(restConfig, "Pilot", revision)
+}
+
+func getIngressGateways(restConfig *rest.Config, revision string) ([]v1.Pod, error) {
+	return getPodsForComponent(restConfig, "IngressGateways", revision)
+}
+
+func getEgressGateways(restConfig *rest.Config, revision string) ([]v1.Pod, error) {
+	return getPodsForComponent(restConfig, "EgressGateways", revision)
+}
+
+func getPodsForComponent(restConfig *rest.Config, component, revision string) ([]v1.Pod, error) {
+	return getPodsWithSelector(restConfig, istioNamespace, &meta_v1.LabelSelector{
+		MatchLabels: map[string]string{
+			label.IstioRev:               revision,
+			label.IstioOperatorComponent: component,
+		},
+	})
+}
+
+func getPodsWithRevision(restConfig *rest.Config, revision string) ([]v1.Pod, error) {
+	return getPodsWithSelector(restConfig, "", &meta_v1.LabelSelector{
+		MatchLabels: map[string]string{
+			label.IstioRev: revision,
+		},
+	})
+}
+
+func getPodsWithSelector(restConfig *rest.Config, ns string, selector *meta_v1.LabelSelector) ([]v1.Pod, error) {
+	client, err := kubernetes.NewForConfig(restConfig)
```

Comment:
i may not be seeing it but don't we need to enumerate and walk all the user namespaces somewhere?

Reply:
Oh! yes. I found it somewhere. I'll change that

Reply:
Passing empty string for namespaces would consider all namespaces. So that shouldn't be a problem?

Reply:
Passing empty string considers all namespaces. So I think it should be fine. See the output of `describe` with `-v` flag. It lists all pods of a particular revision. There is one tiny issue though - it includes istio components like istiod, gateways.